// @parafin/react-parafin-elements/src/deprecated/button/index.tsx
import React from 'react'
import ParafinElements from 'main'
import { ParafinProduct } from 'layers/entry/types'


type ButtonStyle = {
  textColor: string
  backgroundColor: string
  fontSize: string
  height: string
  borderRadius: string
}

type Props = {
  token: string
  environment: 'development' | 'production'
  onExit?: (metadata: any) => void
  onExitV2?: (metadata: any) => void
  style: Partial<ButtonStyle>
  product: ParafinProduct
}
const Button = ({ token, environment, onExit, onExitV2, style, product }: Props) => {
  return (
    <ParafinElements
      token={token}
      getToken={() => Promise.resolve(token)}
      environment={environment}
      onExit={onExit}
      onExitV2={onExitV2}
      variant="button"
      product={product}
      style={{
        variants: {
          button: {
            background: style.backgroundColor,
            color: style.textColor,
            fontSize: style.fontSize,
            height: style.height,
            borderRadius: style.borderRadius,
            borderWidth: 0,
          }
        }
      }}
      overrides={{
        __PRODUCT__: {
          widgetURL: `https://button.widgets${environment === 'production' ? '' : '.dev'}.parafin.com`,
        }
      }}
    />
  )
}

export default Button
