// @parafin/react-parafin-elements/src/layers/entry/types.ts
export type AccountManager = {
  name: string
  email: string
}

type Address = {
  addressLine1: string
  addressLine2?: string
  city: string
  state: string
  postalCode: string
  country: string
}

export type BusinessDetailsV2 = {
  businessExternalId: string
  legalBusinessName: string
  dbaName?: string
  ownerFirstName: string
  ownerLastName: string
  accountManagers?: AccountManager[]
  routingNumber?: string
  accountNumberLastFour: string
  email: string
  phoneNumber: string
  address: Address
}

// deprecated type for maintaining backward compatibility
export type BusinessDetails = {
  businessExternalId: string
  businessName: string
  ownerFirstName: string
  ownerLastName: string
  accountManagers?: AccountManager[]
  routingNumber?: string
  accountNumberLastFour: string
  email: string
  postalCode: string
}

// Any = "merchant_cash_advance" | "flex_loan". We support these for partner dashboard, but don't want to type them for partners.
export type ParafinProduct = 'card' | 'capital' | string
